src/components/colorscale/make_scale_function.js

```diff
@@ -13,35 +13,89 @@ var d3 = require('d3');
 var tinycolor = require('tinycolor2');
 var isNumeric = require('fast-isnumeric');
 
-var Lib = require('../../lib');
 var Color = require('../color');
 
+/**
+ * General colorscale function generator.
+ *
+ * By default, the routine compute that domain and range,
+ * but optional can be called with pre-computed values.
+ *
+ * @param {array} scl
+ *  plotly.js colorscale array of arrays as found in fullData
+ * @param {number} cmin
+ *  minimum color value (used to clamp scale)
+ * @param {number} cmax
+ *  maximum color value (used to clamp scale)
+ * @param {object} opts [optional]
+ *      - domain {array} precomputed domain
+ *      - range {array} precomputed range
+ *      - noNumericCheck {boolean} if true, scale func bypasses numeric checks
+ *      - returnArray {boolean} if true, scale func return 4-item array instead of color strings
+ *
+ * @return {function}
+ *
+ */
+module.exports = function makeScaleFunction(scl, cmin, cmax, opts) {
```

Comment:
this signature is a bit funny, especially when you have calls like
```
makeScaleFunction(scl, null, null, {domain: domain, range: range})
```
where `scl` gets ignored entirely. Maybe better to just put *everything* into opts?